### ISSUE TITLE
No output from backend onChange script

### ISSUE THREAD

**Issue by konobi** [joyent/containerpilot#210]
I've got a project here where I'm having trouble with the onChange script, but containerpilot doesn't seem to pipe the output from this script to it's own stdout/stderr. I have no indication as to wether the script was run, attempted to be run or if it failed for some reason, so I'm kinda stuck =0(

It could be related to the last releases changes, but I have the same happening with older versions of containerpilot. This report is for 2.4.1.

**Comment by fitz123** [comment]
Hello knonobi!
Do you see output if your onChange script is `echo "I'm the stdout Output!"` ?

**Comment by misterbisson** [comment]
You might try changing increasing the log level in the logging config, https://www.joyent.com/containerpilot/docs/configuration#-code-logging--code-

**Comment by konobi** [comment]
nope, and i have `set +x` set within the script, so I should be seeing lots of output. The logging is already set at INFO.

**Comment by misterbisson** [comment]
> The logging is already set at INFO

Have you tried `DEBUG`?

**Comment by konobi** [comment]
ah... not even `echo '==============I AM HERE==============='` defined directly in the .json is being shown... lemme try DEBUG.
K, there some information being displayed now, but I don't see any output from the process, just information about the go functions being used; I'm not seeing the above output in the docker logs =0(
I can't even tell if it's attempting to be run on reload... =0/

**Comment by fitz123** [comment]
I'd guess that your onChange script is not triggered, probably because ether containerpilot have problem to communicate with consul or you're using wrong name in _backend_ section

with log INFO level echos should be visible.

That's my example, try simplify it:

1. Run Container
```
{
  "consul": "{{ .CONSUL }}",
  "logging": {"level": "INFO"},
  "services": [
    {
      "name": "origin-proxy",
      "port": 8888,
      "poll": 2,
      "interfaces": ["10.0.0.0/8"],
      "ttl": 5
    }
  ],
  "backends": [
    {
      "name": "origin-proxy",
      "poll": 1,
      "onChange": "echo OnCHange",
      "timeout": "1s"
    }
  ]
}
```
Immediately after run log:
```
2016/08/17 19:53:36 OnCHange
```
Then manually deregistered service: `curl-XPUT 'localhost:8500/v1/agent/service/deregister/origin-proxy-admin'`

Then logs output:
```
2016/08/17 19:53:57 Unexpected response code: 500 (CheckID does not have associated TTL)
Service not registered, registering...
2016/08/17 19:54:00 OnCHange
```

**Comment by tgross** [comment]
If you can share the debug logs and containerpilot.json config file (redacted if necessary) in a gist that might be helpful for us to help debug this.

**Comment by konobi** [comment]
ah... so does the onChange handler not get called on initial startup or containerpilot reload time?

**Comment by konobi** [comment]
I'm already using the prestart to do a whole bunch of things, it's just that the onChange script that I have is designed purely for making sure the backend service is up and marking something as "primary". Since that needs to occur on any change to that backend service, that's all it does. 

However, since it is never run for any of the instances on start up, none of them will be attempting to claim themselves as primary. When the containerpilot process comes up (or reloads), I was assuming that it was going to fire these scripts since "starting" and "reloading" are states in and of themselves.

I could call the onChange script from prestart I suppose, but it seems kinda odd.

**Comment by tgross** [comment]
Yeah the `onChange` only gets fired when a monitored backend changes. Once the `preStart` exits successfully, the various polling loops start (async) and then the main application is started.

**Comment by konobi** [comment]
ah... i had expected that a "starting" state would have been thought of as an onChange event since it's up to the image to decide which one is primary.

I'm having a little more luck now with trying to run the onChange script at the end of the prestart script, but it gets into nasty issues of daemonizing properly and tty issues, etc. since the onChange script just wants to be spun off initially. I have to push all the output out to /dev/null because otherwise the prestart script doesn't exit until it's file descriptors are cleaned up.

I support you could call it a preexec hook, since it's only interested in being called when the main process is being run, but along side it. Any suggestions on how to achieve the same sort of thing?

**Comment by fitz123** [comment]
@konobi , if I understand your use case correctly:
1. Use consul's tags to mark service as primary/secondary. You have to set `"enableTagOverride": true` in consul to make it work.
2. Mark service _Slave_ by default
3. Use both prestart and onchange hooks to check status of the service
4. Became _Primary_ in case no healthy nodes are detected

**Comment by misterbisson** [comment]
> Use consul's tags to mark service as primary/secondary

I'd suggest using locks, rather than tags. You can see that in use in https://github.com/autopilotpattern/mysql in `bin/manage.py`. Locking semantics are well suited to situations where there can be only one primary.

**Comment by misterbisson** [comment]
> > does the onChange handler not get called on initial startup or containerpilot reload time?

> since it is never run for any of the instances on start up, none of them will [...] When the containerpilot process comes up (or reloads), I was assuming that it was going to fire [the `onChange`] scripts [...]

@fitz123 and @tgross correctly clarified that the `backends` → `onChange` handler _only_ applies to the monitored backend specified in the `name` field. See https://www.joyent.com/containerpilot/docs/configuration#-code-backends--code- for docs on that.

The underlying question here is how to execute code before the main application starts, or before it's registered in the service catalog. The canonical example for simple cases can be found in [`autopilotpattern/nginx`](https://github.com/autopilotpattern/nginx/blob/master/etc/containerpilot.json)'s [`containerpilot.json`](https://github.com/autopilotpattern/nginx/blob/master/etc/containerpilot.json):

```json
{
  "consul": "{{ if .CONSUL_AGENT }}localhost{{ else }}{{ .CONSUL }}{{ end }}:8500",
  "preStart": "/usr/local/bin/reload.sh preStart",
  "logging": {"level": "DEBUG"},
  "services": [
    {
      "name": "nginx",
      "port": 80,
      "health": "/usr/bin/curl --fail --silent --show-error --output /dev/null http://localhost/nginx-health",
      "poll": 10,
      "ttl": 25
    }
  ],
  "backends": [
    {
      "name": "{{ .BACKEND }}",
      "poll": 7,
      "onChange": "/usr/local/bin/reload.sh"
    }
  ],
[...]
```

In that example, Nginx needs to have its configuration template written out with any backends before it starts. The solution there is to use a [`preStart` handler](https://github.com/autopilotpattern/nginx/blob/master/bin/reload.sh#L6-L14) that is almost identical to the [`onChange` handler](https://github.com/autopilotpattern/nginx/blob/master/bin/reload.sh#L16-L24), except that it doesn't attempt to reload Nginx.

MySQL is a more complex application with a suitable more sophisticated set of handlers. Nonetheless, it's still using the normal lifecycle events in ContainerPilot. [The `preStart`](https://github.com/autopilotpattern/mysql/blob/master/etc/containerpilot.json#L3) runs before MySQL starts, but we're also leveraging [`services` → `health`](https://github.com/autopilotpattern/mysql/blob/master/etc/containerpilot.json#L8) and watching for changes to the `mysql-primary` [in `backends` → `onChange`](https://github.com/autopilotpattern/mysql/blob/master/etc/containerpilot.json#L13-L18). As noted in my comment above, `bin/manage.py` in https://github.com/autopilotpattern/mysql may offer a number of solution that you can borrow from.

I'm going to label this as `usage`, and I've added a note in https://github.com/joyent/containerpilot/issues/197#issuecomment-240557097 to improve the docs about this.

**Comment by konobi** [comment]
Okay, I ended up with something working. I ended up making a single non-restarting co-process that calls my onChange script. So the script is kicked off at the right time, ensures that all instances can attempt to become primary all while stateless. The only stateful part is on the primary where it using a consul session lock which is renewed by the health script.

It'd be handy to maybe see a flag on the backends so that on start/reload the onChange script is run (90% of the time it'll just do nothing, but gives it the option of doing something, like becoming the primary).
